KBS URI

Xynnn007 opened the pull request:
As https://github.com/confidential-containers/guest-components/issues/191, this PR includes the following
- Introduce KBS URI scheme
- implement cc-kbc's API
- implement sample-kbc's API
- implement offline-fs-kbc's API
- implement offline-sev-kbc's API
- implement online-sev-kbc's API
- implement eaa-kbc's API
- document about the KBS URI

Please notice that all the `<kbs-addr>` field in a kbs uri now will be ignored, all kbc will use its own member `kbs_url` instead.

Also, this PR modifies the gRPC API of `get_resource`, as duplicated information is carried in the kbs uri

Xynnn007 committed:
feat: kbs uri support for decrypt_payload

key_id in the AnnotationPacket will follow the KBS URI format.
Different KBCs will index the corresponding keu due to the uri.

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: kbs uri support for get_resource

now get_resource API receives a kbs uri to indicate the
confidential resource to be requested

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: eaa-kbc support get_resource kbs uri

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: cc kbc support get_resource kbs uri

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: offline-fs-kbc support get_resource kbs uri

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: online-sev-kbc support get_resource kbs uri

Now if the request resource `type` in kbs uri is
`"client-id"`, the responsed resource will be the
client id.

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
feat: sample kbc support get_resource kbs uri

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
docs: kbs uri guide and specification

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
note: kbs url usage warning in note

please refer to https://github.com/confidential-containers/attestation-agent/issues/130
for more information

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
fix: use wraptype defined in RFC7518

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

jialez0 commented on src/kbc_modules/eaa_kbc/mod.rs:
```diff
@@ -171,12 +177,11 @@ impl EAAKbc {
-        let resource_info = self.kbs_get_resource_info(desc.name.as_str())?;
+    fn kbs_get_resource(&mut self, rid: &ResourceId) -> Result<Vec<u8>> {
+        let resource_path = rid.resource_path();
+        let command = format!("Get {resource_path}");
```
If the command is modified here, `Verdictd` needs to be modified in to match it.

Xynnn007 replied:
Yes, related issue is https://github.com/inclavare-containers/verdictd/issues/29 and we could handle later

jialez0 commented on src/kbc_modules/offline_fs_kbc/common.rs:
```diff
@@ -84,11 +97,11 @@ pub mod tests {
-          ResourceName::GPGPublicKey.to_string(): base64::encode(PUBKEY.as_bytes()),
-          ResourceName::CosignVerificationKey.to_string(): base64::encode(COSIGNKEY.as_bytes()),
-          ResourceName::Credential.to_string(): base64::encode(CREDENTIAL.as_bytes()),
+            resource_path!(ResourcePath::Policy): base64::encode(POLICYJSON.as_bytes()),
```
The value of `resource_path!(ResourcePath::Policy)` is `example.org/test-repo/security-policy/test`, right? Does `example.org` need to be removed?

Xynnn007 replied:
Great, you're right, sorry for the mistake

jialez0 commented on src/kbc_modules/offline_fs_kbc/mod.rs:
```diff
@@ -49,12 +50,15 @@ impl KbcInterface for OfflineFsKbc {
-    async fn get_resource(&mut self, description: &str) -> Result<Vec<u8>> {
-        let desc: ResourceDescription = serde_json::from_str::<ResourceDescription>(description)?;
+    async fn get_resource(&mut self, rid: ResourceId) -> Result<Vec<u8>> {
+        let resource_pahth = rid.resource_path();
```
Nit: `resource_pahth` -> `resource_path`.

jialez0 commented on src/kbc_modules/uri.rs:
```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//! Resource Id is the url of KBS URL.
```
`//! ResourceId is the identification information of all resources that need to be obtained from `get_resource` endpoint.` is better, I think.

jialez0 commented on src/kbc_modules/uri.rs:
```diff
@@ -0,0 +1,166 @@
+
+const SCHEME: &str = "kbs";
+
+/// Resource Id is proposed in <https://github.com/confidential-containers/documentation/issues/85>
```
It is better to link to the KBS URI description document here, not issue.

jialez0 commented on src/kbc_modules/uri.rs:
```diff
@@ -0,0 +1,166 @@
+impl TryFrom<&str> for ResourceId {
+    type Error = &'static str;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
```
Perhaps the implementation of this function can be simplified into the following two lines?
```rust
let url = url::Url::try_from(value)?;
Self::try_from(url)
```

jialez0 reviewed (commented):
@Xynnn007 Thanks a lot!. This is an important PR and plays a great role in our follow-up work. Here are some small questions and comments...

BTW, the changes in this PR will make the current `sample_keyprovider` unusable. Is there a plan to modify the `sample_keyprovider`?

Xynnn007 commented on docs/KBS_URI.md:
```diff
@@ -0,0 +1,33 @@
+
+### CC-KBC
+
+CC-KBC will convert a KBS URI into a web link to request a coco-KBS. The web link follows the [OpenAPI of KBS](https://github.com/confidential-containers/kbs/blob/main/docs/kbs.yaml#L74).
```
Great suggestion!

Xynnn007 commented:
@jialez0 Thanks a lot for deep diving... I will update the keyprovider and related documents in the next PR. For sample-key provider, there might be two ways:
- give the new key provider a parameter to determine whether to use the hardcoded key of sample kbc
- deprecate it

I prefer way 2, some related work has already been done in image-rs https://github.com/confidential-containers/image-rs/issues/85 . And we can get rid of those encrypted images using sample-kbc then (if any). Finally we can mark sample kbc deprecated in code and give a compiling warning or directly deleted the related code

Xynnn007 commented:
Hi @stevenhorsman @liudalibj this pr will influence the behavior of `offline-kbs` in resource indexing and we will need to remake some test images for kata-cc test. Please take a look if convenient to make sure I did not do anything stupid : )

Xynnn007 commented on protos/getresource.proto:
```diff
@@ -3,9 +3,8 @@ syntax = "proto3";
-    string KbcName = 1;
-    string KbsUri = 2;
-    string ResourceDescription = 3;
+    string ResourceUri = 1;
```
I changed the Proto here, as kbsUri is part of the resourceUri. KbcName now still matters and I think we can open an issuer later to talk about https://github.com/confidential-containers/attestation-agent/issues/99 on both image-rs side and kata-agent side.

liudalibj commented on src/kbc_modules/offline_fs_kbc/mod.rs:
```diff
@@ -102,111 +104,82 @@ mod tests {
+    #[case(true, ResourcePath::CosignVerificationKey.as_ref(), COSIGNKEY)]
+    #[case(true, ResourcePath::Credential.as_ref(), CREDENTIAL)]
+    // Case 2. Error while get bad resource name from a good kbc instance
+    #[case(false, "bad", "")]
```
I meet a test error at here, maybe the bad cases need be updated, would you like do a check?
```bash
root@peer-pod1:~/attestation-agent# make KBC=offline_fs_kbc && make install
root@peer-pod1:~/attestation-agent# cargo test --features offline_fs_kbc,rust-crypto
warning: skipping duplicate package `app` found at `/root/.cargo/git/checkouts/sgxdatacenterattestationprimitives-d6934a418e6beae0/85cf8bd/SampleCode/RustTDQuoteGenerationSample`
    Finished test [unoptimized + debuginfo] target(s) in 0.14s
     Running unittests src/lib.rs (target/debug/deps/attestation_agent-c31f2de2717cdaeb)

running 12 tests
test kbc_modules::offline_fs_kbc::tests::test_get_key ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_1 ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_2 ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_3 ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_4 ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_5 ... ok
test kbc_modules::offline_fs_kbc::tests::test_get_resource::case_6 ... FAILED
test kbc_modules::uri::tests::conversions ... ok
test kbc_modules::uri::tests::serialize ... ok
test kbc_modules::uri::tests::deserialize ... ok
test kbc_modules::offline_fs_kbc::common::tests::test_load_keys ... ok
test kbc_modules::offline_fs_kbc::common::tests::test_load_resources ... ok

failures:

---- kbc_modules::offline_fs_kbc::tests::test_get_resource::case_6 stdout ----
thread 'kbc_modules::offline_fs_kbc::tests::test_get_resource::case_6' panicked at 'called `Result::unwrap()` on an `Err` value: "invalid kbs uri, should be kbs://<addr-of-kbs>/<repo>/<type>/<tag>"', src/kbc_modules/offline_fs_kbc/mod.rs:175:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    kbc_modules::offline_fs_kbc::tests::test_get_resource::case_6

test result: FAILED. 11 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

Xynnn007 replied:
Thanks for pointing out! This is caused by the KBS URI parsed error. I've fixed this. There are also no related CI to cover. I've updated `basic.yaml` CI to cover this.

liudalibj reviewed (changes requested):
one offline_fs_kbc test case need be updated

Xynnn007 committed:
ci: let basic ci cover offline-fs-kbc

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

liudalibj reviewed (approved):
Thanks for the quick action, the whole PR looks good to me.

fitzthum reviewed (commented):
A few initial comments.

Xynnn007 commented on src/kbc_modules/online_sev_kbc/mod.rs:
```diff
@@ -49,7 +49,9 @@ impl KbcInterface for OnlineSevKbc {
     async fn decrypt_payload(&mut self, annotation_packet: AnnotationPacket) -> Result<Vec<u8>> {
-        let key = self.get_key_from_kbs(annotation_packet.kid).await?;
+        let key = self
+            .get_key_from_kbs(annotation_packet.kid.resource_path())
```
Yes, we are. In online-sev-kbc, it also defines its own kbs address like cc-kbc and the `get_key_from_kbs` only makes a gRPC request containing the key id to kbs. So here we only need to extract the `<repository>/<type>/<tag>` (what `resource_path()` does) and ignore the uri of kbs. (like `get_resource_from_kbs` does)

Also, `get_resource_from_kbs` is a little different from `get_key_from_kbs`. The later will flush the memory after using the key using `Zeroizing`

Xynnn007 commented on src/kbc_modules/uri.rs:
```diff
@@ -0,0 +1,154 @@
+const RESOURCE_ID_ERROR_INFO: &str =
+    "invalid kbs uri, should be kbs://<addr-of-kbs>/<repo>/<type>/<tag>";
+
+const SCHEME: &str = "kbs";
```
I think yes. For example the `policy.json` in image-rs may look like
```json
{
    "default": [{"type": "reject"}],
    "transports": {
        "docker": {
            "docker.io/my_private_registry": [
                {
                    "type": "sigstoreSigned",
                    "keyData": "<public Key>",
                }
            ],
            "registry.access.redhat.com": [
                {
                    "type": "sigstoreSigned",
                    "keyPath": "/etc/key1",
                }
            ],
            "quay.io/confidential-containers": [
                {
                    "type": "sigstoreSigned",
                    "keyPath": "kbs://example.org/repo/typ/tag",
                }
            ],
        },
        "dir": {
            "": [{"type": "insecureAcceptAnything"}]
        }
    }
}
```

Which will help to identify `local fs`, `kbs resource`, and potential `http`, ...

Xynnn007 commented on src/kbc_modules/sample_kbc/mod.rs:
```diff
@@ -4,14 +4,16 @@
 use std::str::FromStr;
 use zeroize::Zeroizing;
 
+use super::tests::ResourcePath;
```
Yes! I'll fix this, thanks!

Xynnn007 commented on src/kbc_modules/sample_kbc/mod.rs:
```diff
@@ -45,27 +47,18 @@ impl KbcInterface for SampleKbc {
+    async fn get_resource(&mut self, rid: ResourceId) -> Result<Vec<u8>> {
+        let uri = rid.whole_uri();
+        match ResourcePath::from_str(&uri).context("Unknown resource name")? {
+            ResourcePath::Policy => Ok(std::include_str!("policy.json").as_bytes().to_vec()),
```
Yes, great suggestion. Thanks!

Xynnn007 commented on src/kbc_modules/offline_fs_kbc/common.rs:
```diff
@@ -62,6 +63,18 @@ pub mod tests {
     #[allow(dead_code)]
     pub const RESOURCES_NAME: &str = "aa-offline_fs_kbc-resources.json";
 
+    pub const KBS_URI_PREFIX: &str = "kbs://example.org/";
```
Yes, thanks! let me fix this.

jialez0 commented on docs/KBS_URI.md:
```diff
@@ -0,0 +1,33 @@
+
+## Introduction
+
+To uniquely identify every resource/key in the CoCo Key Broker System, a __KBS URI__ is defined.
```
Why don't we call such URI `Resource URI`? If we use the name `KBS URI` to refer to a URI such as "kbs://", it will be easy to confuse with the "https://" URI of the KBS instance.

fitzthum replied:
Agree with @jialez0. KBS URI sounds like it points to a KBS not a resource.

Xynnn007 replied:
Great, nice suggestion. Thanks!

Xynnn007 committed:
fix: rename kbs uri to kbs resource uri

To avoid confusing, a uri pointing to a resource
is named kbs resource uri. A kbs uri stands for
the uri of the kbs.

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 commented:
Hi @fitzthum @jialez0 I've changed the name `ResourceId` into `ResourceUri`. Please take another look to make sure I didnot do anything wrong. Thanks!

jialez0 commented on src/kbc_modules/uri.rs:
```diff
@@ -0,0 +1,154 @@
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+const RESOURCE_ID_ERROR_INFO: &str =
+    "invalid kbs uri, should be kbs://<addr-of-kbs>/<repo>/<type>/<tag>";
```
`invalid kbs uri` -> `invalid kbs resource uri` is better, I think.

Xynnn007 replied:
Thanks, fixed it !

jialez0 reviewed (approved):
Note: To be compatible with this PR, `image-rs` may need some updates and its existing resource acquisition format will no longer be available, right? @Xynnn007

Xynnn007 commented:
> Note: To be compatible with this PR, `image-rs` may need some updates and its existing resource acquisition format will no longer be available, right? @Xynnn007

Yes, we need to remove the hardcodes in image-rs and give a configuration.

Xynnn007 commented:
There are some conflicts in the code for ttrpc things.. let me finish some tiny fixes and then rebase the code.

Xynnn007 committed:
fix: delete lib Cargo.lock file

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

Xynnn007 committed:
App: apply the new get_resource API

Signed-off-by: Xynnn007 <xynnn@linux.alibaba.com>

jialez0 commented:
> Yes, we need to remove the hardcodes in image-rs and give a configuration of image-rs for the resources needed. This work should be delayed after 0.4.0 is released. Let's perfect the system for 0.5.0 : )

@Xynnn007 Sounds great! Since the changes in this PR will affect the code availability of `image-rs` and `sample keyprovider`, we should wait for the above two adaptations are ready, then merge the PR of `AA`, `image-rs` and `sample keyprovider` at the same time, so that we can ensure that the existing upstream code is available as much as possible.

cc @fitzthum

jialez0 commented:
@Xynnn007 In view of the fact that the KBS address part of the KBS URI is not actually consumed at present, we actually use the KBS address specified in the `aa_kbc_params` field in the config of kata-agent. Therefore, in the current version, I suggest that we temporarily directly set the format of the KBS URI to:
```
kbs:///<repository>/<type>/<tag>
```
This can avoid confusion for users now.

Xynnn007 commented:
> @Xynnn007 In view of the fact that the KBS address part of the KBS URI is not actually consumed at present, we actually use the KBS address specified in the `aa_kbc_params` field in the config of kata-agent. Therefore, in the current version, I suggest that we temporarily directly set the format of the KBS URI to:
> 
> ```
> kbs:///<repository>/<type>/<tag>
> ```
> 
> This can avoid confusion for users now.

Yea, I have no preference for either option `kbs:///<repository>/<type>/<tag>` or `kbs://<kbs-addr>/<repository>/<type>/<tag>`. I just want to say that `<kbs-addr>` could be used in future potentially to distinguish resources from different KBS. Now`kbs:///<repository>/<type>/<tag>` presentation may in the following scenarios
- Test uris in `kbc_modules::tests::ResourcePath`
- Documents
- Guide for CoCo Key Provider

Do they all need change?

jialez0 commented:
> Nowkbs:///<repository>/<type>/<tag> presentation may influence the following scenarios

> - Test uris in kbc_modules::tests::ResourcePath
> - Documents
> - Guide for CoCo Key Provider
> Does it look good?

@Xynnn007 That's OK. In fact, whether we use `example.org` or null, they will not be actually read under the current system. I think it would be better to use null values at present, so I suggest to implement this modification.

Meanwhile, in `coco keyprovider` and `image-rs`, where `example.org` is used, I think null values can be used instead

fitzthum reviewed (approved):
LGTM. I also prefer the adjustment that @jialez0 mentions regarding the URI, but we can do that later.
